services/local-notifications/core.go

```diff
@@ -0,0 +1,338 @@
+	} else if to != nil {
+		deeplink = walletDeeplinkPrefix + to.Address.String()
+	}
+
+	body := notificationBody{
+		State:       state,
+		From:        transfer.From,
+		To:          transfer.Address,
+		FromAccount: from,
+		ToAccount:   to,
+		Value:       transfer.Value,
+		ERC20:       string(transfer.Type) == "erc20",
+		Contract:    transfer.Contract,
+		Network:     transfer.NetworkID,
+	}
+
+	return &Notification{
+		ID:       transfer.ID,
+		Body:     body,
+		Deeplink: deeplink,
+		Category: "transaction",
+	}
+}
+
+func (s *Service) transactionsHandler(payload TransactionEvent) {
+	log.Info("Handled a new transaction", "info", payload)
+
+	limit := 20
+	if payload.BlockNumber != nil {
+		for _, address := range payload.Accounts {
+			log.Info("Handled transfer for address", "info", address)
+			transfers, err := s.walletDB.GetTransfersByAddressAndBlock(address, payload.BlockNumber, int64(limit))
+			if err != nil {
+				log.Error("Could not fetch transfers", "error", err)
+			}
+
+			for _, transaction := range transfers {
+				n := s.buildTransactionNotification(transaction)
+				pushMessage(n)
+			}
+		}
+	}
+}
+
+// SubscribeWallet - Subscribes to wallet signals
+func (s *Service) SubscribeWallet(publisher *event.Feed) error {
+	s.walletTransmitter.publisher = publisher
+
+	preference, err := s.db.GetWalletPreference()
+
+	if err != nil {
+		log.Error("Failed to get wallet preference", "error", err)
+		return nil
+	}
+
+	if preference.Enabled {
+		s.StartWalletWatcher()
+	}
+
+	return nil
+}
+
+// StartWalletWatcher - Forward wallet events to notifications
+func (s *Service) StartWalletWatcher() {
+	if s.walletTransmitter.quit != nil {
+		// already running, nothing to do
+		return
+	}
+
+	if s.walletTransmitter.publisher == nil {
+		log.Error("wallet publisher was not initialized")
+		return
+	}
+
+	s.walletTransmitter.quit = make(chan struct{})
+	events := make(chan wallet.Event, 10)
+	sub := s.walletTransmitter.publisher.Subscribe(events)
+
+	s.walletTransmitter.wg.Add(1)
+
+	go func() {
+		defer s.walletTransmitter.wg.Done()
+		for {
+			select {
+			case <-s.walletTransmitter.quit:
+				sub.Unsubscribe()
+				return
+			case err := <-sub.Err():
+				// technically event.Feed cannot send an error to subscription.Err channel.
+				// the only time we will get an event is when that channel is closed.
+				if err != nil {
+					log.Error("wallet signals transmitter failed with", "error", err)
+				}
+				return
+			case event := <-events:
+
+				if event.Type == wallet.EventNewBlock {
+					s.transmitter.publisher.Send(TransactionEvent{
+						Type:                      string(event.Type),
+						BlockNumber:               event.BlockNumber,
+						Accounts:                  []common.Address(event.Accounts),
+						NewTransactionsPerAccount: map[common.Address]int(event.NewTransactionsPerAccount),
+						ERC20:                     bool(event.ERC20),
+					})
+				}
+			}
+		}
+	}()
```

Comment:
What is the functional difference between this implementation and the below implementation?

```suggestion
	for {
		select {
		case <-s.walletTransmitter.quit:
			sub.Unsubscribe()
			return
		case err := <-sub.Err():
			// technically event.Feed cannot send an error to subscription.Err channel.
			// the only time we will get an event is when that channel is closed.
			if err != nil {
				log.Error("wallet signals transmitter failed with", "error", err)
			}
			return
		case event := <-events:
				if event.Type == wallet.EventNewBlock {
				s.transmitter.publisher.Send(TransactionEvent{
					Type:                      string(event.Type),
					BlockNumber:               event.BlockNumber,
					Accounts:                  []common.Address(event.Accounts),
					NewTransactionsPerAccount: map[common.Address]int(event.NewTransactionsPerAccount),
					ERC20:                     bool(event.ERC20),
				})
			}
		}
	}
```

We start a subroutine but block the parent function by adding to the wait group.

So either we don't need to wait and run a routine, or we do need the routine but we don't need to wait.